pkg/controller/failurerecovery/pod_termination_controller.go

```diff
@@ -0,0 +1,199 @@
+
+const (
+	KueueFailureRecoveryConditionType = "KueueFailureRecovery"
+	KueueForcefulTerminationReason    = "KueueForcefullyTerminated"
+)
+
+type TerminatingPodReconciler struct {
+	client                         client.Client
+	clock                          clock.Clock
+	forcefulTerminationGracePeriod time.Duration
+	recorder                       record.EventRecorder
+}
+
+type TerminatingPodReconcilerOptions struct {
+	clock                          clock.Clock
+	forcefulTerminationGracePeriod time.Duration
+}
+
+type TerminatingPodReconcilerOption func(*TerminatingPodReconcilerOptions)
+
+func WithClock(c clock.Clock) TerminatingPodReconcilerOption {
+	return func(o *TerminatingPodReconcilerOptions) {
+		o.clock = c
+	}
+}
+
+func WithForcefulTerminationGracePeriod(t time.Duration) TerminatingPodReconcilerOption {
+	return func(o *TerminatingPodReconcilerOptions) {
+		o.forcefulTerminationGracePeriod = t
+	}
+}
+
+var defaultOptions = TerminatingPodReconcilerOptions{
+	clock:                          realClock,
+	forcefulTerminationGracePeriod: time.Minute,
+}
+
+func NewTerminatingPodReconciler(
+	client client.Client,
+	recorder record.EventRecorder,
+	opts ...TerminatingPodReconcilerOption,
+) (*TerminatingPodReconciler, error) {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &TerminatingPodReconciler{
+		client:                         client,
+		clock:                          options.clock,
+		forcefulTerminationGracePeriod: options.forcefulTerminationGracePeriod,
+		recorder:                       recorder,
+	}, nil
+}
+
+func (r *TerminatingPodReconciler) Generic(event.GenericEvent) bool {
+	return false
+}
+
+func (r *TerminatingPodReconciler) Create(e event.CreateEvent) bool {
+	pod := e.Object.(*corev1.Pod)
+
+	if !podOptedInToFailurePolicy(pod) {
+		return false
+	}
+
+	if pod.DeletionTimestamp.IsZero() {
+		return false
+	}
+
+	return true
+}
+
+func (r *TerminatingPodReconciler) Update(u event.UpdateEvent) bool {
+	oldPod := u.ObjectOld.(*corev1.Pod)
+	newPod := u.ObjectNew.(*corev1.Pod)
+
+	if !podOptedInToFailurePolicy(newPod) {
+		return false
+	}
+
+	// Pod was not marked for deletion in the update
+	if !oldPod.DeletionTimestamp.IsZero() || newPod.DeletionTimestamp.IsZero() {
+		return false
+	}
+
+	return true
+}
+
+func (r *TerminatingPodReconciler) Delete(event.DeleteEvent) bool {
+	return false
+}
+
+func podOptedInToFailurePolicy(p *corev1.Pod) bool {
+	annotationValue, hasAnnotation := p.Annotations[constants.SafeToForcefullyTerminateAnnotationKey]
+	return hasAnnotation && annotationValue == constants.SafeToForcefullyTerminateAnnotationValue
+}
+
+func (r *TerminatingPodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	pod := &corev1.Pod{}
+	if err := r.client.Get(ctx, req.NamespacedName, pod); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Pod was already terminated
+	if utilpod.IsTerminated(pod) {
```

Comment:
Add the check to the event handlers too. And wrap all the Pod related checks to a helper
`func preconditionsMet(p *corev1.Pod) bool`, feel free to change the naming.
Then from Update call it for the `newPod`